pyHalo/preset_models.py

```diff
@@ -239,4 +241,122 @@ def SIDM(z_lens, z_source, cross_section_name, cross_section_class, kwargs_cross
+    :param log_m_host: log10 host halo mass in M_sun
+    :param power_law_index: logarithmic slope of the subhalo mass function
+    :param r_tidal: subhalos are distributed following a cored NFW profile with a core radius r_tidal. This is intended
+    to account for tidal stripping of halos that pass close to the central galaxy
+    :param mass_definition: mass profile model for halos
+    :param log10_m_uldm: ULDM particle mass in log units, typically 1e-22 eV
+    :param uldm_plaw: ULDM core radius-halo mass power law exponent, typically 1/3
+    :param kwargs_other: any other optional keyword arguments
+    :return: a realization of ULDM halos
+    """
+    # constants
+    m22 = 10**(log10_m_uldm + 22)
+    log_m0 = np.log10(1.6e10 * m22**(-4/3))
+    M_min0 = 4.4e7 * m22**(-3/2) # M_solar
+
+    a_uldm = 1 # set to unity since Schive et al. 2016 do not have an analogous parameter
+
+    #compute M_min as described in documentation
+    a = lambda z: (1+z)**(-1)
+    O_m = lambda z: Cosmology().astropy.Om(z)
+    zeta = lambda z: (18*np.pi**2 + 82*(O_m(z)-1) - 39*(O_m(z)-1)**2) / O_m(z)
+    m_min = lambda z: a(z)**(-3/4) * (zeta(z)/zeta(0))**(1/4) * M_min0
+    log_mlow = lambda z: np.log10(m_min(z))
+
+    kwargs_model_field = {'a_wdm': a_uldm, 'b_wdm': b_uldm, 'c_wdm': c_uldm, 'log_mc': log_m0,
+                          'c_scale': c_scale, 'c_power': c_power, 'log_mlow': log_mlow, 'log_mhigh': log_mhigh,
+                          'cone_opening_angle': cone_opening_angle_arcsec, 'mdef_los': mass_definition,
+                          'mass_func_type': 'POWER_LAW', 'LOS_normalization': LOS_normalization, 'log_m_host': log_m_host,
+                          }
+
+    kwargs_model_subhalos = {'a_wdm': a_uldm, 'b_wdm': b_uldm, 'c_wdm': c_uldm, 'log_mc': log_m0,
+                          'c_scale': c_scale, 'c_power': c_power, 'log_mlow': log_mlow, 'log_mhigh': log_mhigh,
+                          'cone_opening_angle': cone_opening_angle_arcsec, 'sigma_sub': sigma_sub, 'mdef_subs': mass_definition,
+                             'mass_func_type': 'POWER_LAW', 'power_law_index': power_law_index, 'r_tidal': r_tidal}
+    
+    kwargs_uldm = {'log10_m_uldm': log10_m_uldm, 'uldm_plaw': uldm_plaw}
+
+    kwargs_model_field.update(kwargs_uldm)
+    kwargs_model_subhalos.update(kwargs_uldm)
+
+    kwargs_model_field.update(kwargs_other)
+    kwargs_model_subhalos.update(kwargs_other)
+
+    # this will use the default cosmology. parameters can be found in defaults.py
+    pyhalo = pyHalo(z_lens, z_source)
+    # Using the render method will result a list of realizations
+    realization_subs = pyhalo.render(['SUBHALOS'], kwargs_model_subhalos, nrealizations=1)[0]
+    realization_line_of_sight = pyhalo.render(['LINE_OF_SIGHT', 'TWO_HALO'], kwargs_model_field, nrealizations=1)[0]
+
+    uldm_realization = realization_line_of_sight.join(realization_subs, join_rendering_classes=True)
+
+    return uldm_realization
```

Comment:
can you add a test function for this function inside the test_preset_models file to make just it executes without throwing an error?

Reply:
Fixed

Reply:
Never mind, called the function incorrectly, will fix shortly.